lib/internal/fs/promises.js

```diff
@@ -739,6 +747,68 @@ async function readFile(path, options) {
   return PromisePrototypeFinally(readFileHandle(fd, options), fd.close);
 }
 
+async function stream(handle, dest, options) {
+  checkAborted(options.signal);
+  if (!EE) {
+    EE = require('events');
+  }
+  const BUFFER_SIZE = 128 * 2 ** 10;
+  const {
+    start = 0,
+    end,
+  } = options;
+
+  if (start && typeof start !== 'number') {
+    throw new ERR_INVALID_ARG_TYPE('start', 'number', start);
+  }
+  if (end && typeof end !== 'number') {
+    throw new ERR_INVALID_ARG_TYPE('end', 'number', end);
+  }
+  if (start < 0) {
+    throw new ERR_OUT_OF_RANGE('start', 'greater than 0', start);
+  }
+  if (end < 0) {
+    throw new ERR_OUT_OF_RANGE('end', 'greater than 0', end);
+  }
+  if (end && start > end) {
+    throw new ERR_OUT_OF_RANGE('start', `less than end ${end}`, start);
+  }
+
+  let currPos = start;
+  const allocatedBuffer = Buffer.allocUnsafe(
+    end ? MathMin(end - currPos, BUFFER_SIZE) : BUFFER_SIZE
+  );
+
+  while (true) {
+    if (end && currPos >= end) {
+      break;
+    }
+
+    checkAborted(options.signal);
+    if (dest.writableNeedDrain) {
+      await EE.once(dest, 'drain');
+    }
+
+    const { buffer, bytesRead } = await handle.read(
+      allocatedBuffer,
+      0,
+      allocatedBuffer.byteLength,
+      currPos
+    );
+
+    if (dest.destroyed) {
+      throw new ERR_STREAM_PREMATURE_CLOSE();
+    }
+
+    if (bytesRead === 0) {
+      break;
+    }
+
+    dest.write(buffer.slice(0, bytesRead));
```

Comment:
This does not appear to be handling backpressure from the downstream writable.

Reply:
It does, here https://github.com/nodejs/node/pull/38440/files#diff-408a6d8eb0a23c586754d4f3af46c8b1927a32d80a6bdfd3bbd87c0f2a012b1aR788-R790